server/repository/src/db_diesel/assets/asset.rs

```diff
@@ -0,0 +1,287 @@
+            match sort.key {
+                AssetSortField::Name => {
+                    apply_sort_no_case!(query, sort, asset_dsl::name);
+                }
+                AssetSortField::SerialNumber => {
+                    apply_sort_no_case!(query, sort, asset_dsl::serial_number);
+                }
+                AssetSortField::InstallationDate => {
+                    apply_sort!(query, sort, asset_dsl::installation_date)
+                }
+                AssetSortField::ReplacementDate => {
+                    apply_sort!(query, sort, asset_dsl::replacement_date)
+                }
+                AssetSortField::ModifiedDatetime => {
+                    apply_sort!(query, sort, asset_dsl::modified_datetime)
+                }
+            }
+        } else {
+            query = query.order(asset_dsl::id.asc())
+        }
+
+        let final_query = query
+            .offset(pagination.offset as i64)
+            .limit(pagination.limit as i64);
+
+        // Debug diesel query
+        // println!(
+        //    "{}",
+        //     diesel::debug_query::<DBType, _>(&final_query).to_string()
+        // );
+
+        let result = final_query.load::<Asset>(&self.connection.connection)?;
+
+        Ok(result.into_iter().map(to_domain).collect())
+    }
+}
+
+fn to_domain(asset_row: AssetRow) -> Asset {
+    asset_row
+}
+
+type BoxedAssetQuery = IntoBoxed<'static, asset::table, DBType>;
+
+fn create_filtered_query(filter: Option<AssetFilter>) -> BoxedAssetQuery {
+    let mut query = asset_dsl::asset.into_boxed();
+
+    if let Some(f) = filter {
+        let AssetFilter {
+            id,
+            name,
+            serial_number,
+            class_id,
+            category_id,
+            type_id,
+            catalogue_item_id,
+            installation_date,
+            replacement_date,
+        } = f;
+
+        apply_equal_filter!(query, id, asset_dsl::id);
+        apply_string_filter!(query, name, asset_dsl::name);
+        apply_string_filter!(query, serial_number, asset_dsl::serial_number);
+
+        apply_equal_filter!(query, catalogue_item_id, asset_dsl::asset_catalogue_item_id);
+        apply_date_filter!(query, installation_date, asset_dsl::installation_date);
+        apply_date_filter!(query, replacement_date, asset_dsl::replacement_date);
+
+        if let Some(category_id) = category_id {
+            let mut sub_query = asset_catalogue_item_dsl::asset_catalogue_item
+                .select(asset_catalogue_item_dsl::id.nullable())
+                .into_boxed();
+            apply_equal_filter!(
+                sub_query,
+                Some(category_id),
+                asset_catalogue_item_dsl::asset_category_id
+            );
+            query = query.filter(asset_dsl::asset_catalogue_item_id.eq_any(sub_query));
+        }
+
+        if let Some(class_id) = class_id {
+            let mut sub_query = asset_catalogue_item_dsl::asset_catalogue_item
```

Comment:
@mark-prins this subquery option is another approach to do what you've done previously.
Should be slightly better performance as a sub query and means you aren't mixing the location where we're handling the query filters. Downside is diesel doesn't make it easy to get right...